src/platform/darwin_openssl.c

```diff
@@ -0,0 +1,195 @@
+/*++
+
+    Copyright (c) Microsoft Corporation.
+    Licensed under the MIT License.
+
+Abstract:
+
+    Implements the certificate functions by calling the darwin keychain store
+
+Environment:
+
+    Darwin
+
+--*/
+
+
+#include "platform_internal.h"
+
+#define OPENSSL_SUPPRESS_DEPRECATED 1 // For hmac.h, which was deprecated in 3.0
+#include "openssl/err.h"
+#include "openssl/hmac.h"
+#include "openssl/kdf.h"
+#include "openssl/pem.h"
+#include "openssl/rsa.h"
+#include "openssl/ssl.h"
+#include "openssl/x509.h"
+
+#include <Security/Security.h>
+
+#ifdef QUIC_CLOG
+#include "darwin_openssl.c.clog.h"
+#endif
+
+BOOLEAN
+CxPlatTlsVerifyCertificate(
+    _In_ X509* X509Cert,
+    _In_ const char* SNI,
+    _In_ QUIC_CREDENTIAL_FLAGS CredFlags
+    )
+{
+    BOOLEAN Result = FALSE;
+    unsigned char* OpenSSLCertBuffer = NULL;
+    CFDataRef CfData = NULL;
+    SecCertificateRef Certificate = NULL;
+    OSStatus Status = 0;
+    CFMutableArrayRef PolicyArray = NULL;
+    SecTrustRef TrustRef = NULL;
+    CFStringRef SNIString = NULL;
+    SecPolicyRef SSLPolicy = NULL;
+    SecPolicyRef RevocationPolicy = NULL;
+    int OpenSSLCertLength = 0;
+
+    OpenSSLCertLength = i2d_X509(X509Cert, &OpenSSLCertBuffer);
```

Comment:
Is this certificate received over the wire? is `i2d` hardened against untrusted input?

Reply:
Yes, this is over the wire. However, since X509Cert is typed coming out of openssl, I'm assuming both its been prevalidated and i2d_X509 is hardened against bad input. We do the same thing on windows currently too.